tests/test_estimation/test_calculate_weights.py

```diff
@@ -0,0 +1,454 @@
+"""Tests for functions in calculate_weights"""
+
+import pandas as pd
+import numpy as np
+import logging
+import pytest
+import src.estimation.calculate_weights as calw
+from pandas._testing import assert_frame_equal, assert_series_equal
+
+
+# Four tests for calc_lower_n:
+# testing calc_lower_n where duplicate refs
+# testing calc_lower_n where missing col
+# testing calc_lower_n where nan in reference
+# testing calc_lower_n where nan in exp_col
+
+
+def test_calc_lower_n():
+    """Test for calc_lower_n with duplicate refs."""
+
+    input_cols = [
+        "reference",
+        "709",
+    ]
+
+    data = [
+        [1, 'A'],
+        [2, 'B'],
+        [2, 'C'],
+        [4, 'D'],
+        [1, 'E'],
+    ]
+
+    input_df = pd.DataFrame(data=data, columns=input_cols)
+
+    # Call calc_lower_n function
+    actual_result = calw.calc_lower_n(input_df)
+
+    # Defined expected result
+    expected_result = 3
+
+    assert actual_result == expected_result
+
+
+def test_calc_lower_n_missing_col():
+    """Test for calc_lower_n with missing col"""
+
+    input_cols = [
+        "ref",
+        "709",
+    ]
+
+    data = [
+        [1, 'A'],
+        [2, 'B'],
+        [2, 'C'],
+        [4, 'D'],
+        [1, 'E'],
+    ]
+
+    input_df = pd.DataFrame(data=data, columns=input_cols)
+
+
+    with pytest.raises(ValueError):
+        calw.calc_lower_n(input_df)
```

Comment:
Can this be amended to account for both cases? Along with message checking for pytest.raises.
Here is an (OPTINONAL) implementation:
```python
@pytest.parameterize(
    "cols, msg, exp_col",
    (
        [["ref", "709"], f"'reference' or 709 missing.", "709"],
        [["reference", "706"], f"'reference' or 707 missing.", "707"]
    )
)
def test_calc_lower_n_missing_col(cols, msg, exp_col):
    """Test for calc_lower_n with missing col"""
    data = [
        [1, 'A'],
        [2, 'B'],
        [2, 'C'],
        [4, 'D'],
        [1, 'E'],
    ]
    input_df = pd.DataFrame(data=data, columns=cols)

    with pytest.raises(ValueError, msg=msg):
        calw.calc_lower_n(input_df, exp_col)
```
NOTE: The above code is untested.